test/test_ourdna_dashboard.py

```diff
@@ -0,0 +1,412 @@
+                samples_filtered.append(sample)
+
+                # Check that the time difference matches
+                assert isinstance(sample.id, str)
+                assert (
+                    time_difference.total_seconds()
+                    == collection_to_process_end_time_24h[sample.id]
+                )
+
+                # Check that the sample id is in the dict
+                assert sample.id in collection_to_process_end_time_24h
+
+        # check that there are a correct number of matching results
+        assert len(collection_to_process_end_time_24h.keys()) == len(samples_filtered)
+
+    @run_as_sync
+    async def test_processing_times_per_site(self):
+        """I want to know what the sample processing times were for samples at each designated site"""
+        dashboard = await self.odd.query(project_id=self.project_id)
+        processing_times_by_site = dashboard.processing_times_by_site
+
+        # Check that processing_times_per_site is not empty and is a dict
+        assert processing_times_by_site
+        assert isinstance(processing_times_by_site, dict)
+
+        sample_tally: dict[str, dict[int, int]] = OrderedDict()
+        for sample in self.sample_external_objects:
+            assert isinstance(sample.meta, dict)
+            processing_site = sample.meta.get('processing-site', 'Unknown')
+            process_start_time = sample.meta.get('process-start-time')
+            process_end_time = sample.meta.get('process-end-time')
+            # Skip samples that don't have process_start_time or process_end_time
+            if not process_start_time or not process_end_time:
+                continue
+            time_difference = str_to_datetime(process_end_time) - str_to_datetime(
+                process_start_time
+            )
+            current_bucket = ceil(time_difference.total_seconds() / 3600)
+            if processing_site in sample_tally:
+                sample_tally[processing_site][current_bucket] += 1
+            else:
+                sample_tally[processing_site] = {}
+                sample_tally[processing_site][current_bucket] = 1
+
+        # Checks that we have identical dicts (by extension, keys and their values)
+        assert OrderedDict(processing_times_by_site) == sample_tally
+
+    @run_as_sync
+    async def test_total_samples_by_collection_event_name(self):
+        """I want to know how many samples were collected from walk-ins vs during events or scheduled activities"""
+        dashboard = await self.odd.query(project_id=self.project_id)
+        total_samples_by_collection_event_name = (
+            dashboard.total_samples_by_collection_event_name
+        )
+
+        # Check that total_samples_by_collection_event_name is not empty and is a dict
+        assert total_samples_by_collection_event_name
+        assert isinstance(total_samples_by_collection_event_name, dict)
+
+        sample_tally: dict[str, int] = OrderedDict()
```

Comment:
This is a test, so I don't thing it matters too much, but I think you could simplify this by using a `defaultdict` here, and then using unittests `self.assertDictEqual`, rather than a straight assert.

In tests, use assert when you want to do something with the typing, otherwise generally encourage using the [unittests methods](https://docs.python.org/3/library/unittest.html#assert-methods) as it provides pretty good support for telling you the value, vs differences are.

Reply:
Great suggestion! I've ported most of the references where I used `assert` except where I need it for typing.